private/bufpkg/bufimage/bufimagemodify/internal/internal.go

```diff
@@ -83,8 +81,18 @@ func RemoveLocationsFromSourceCodeInfo(sourceCodeInfo *descriptorpb.SourceCodeIn
 	// are structured as expected, and collect all of the indices that
 	// we need to delete.
 	indices := make(map[int]struct{}, len(paths)*2)
+	// each path in this trie is for a FieldOptions message (not for a singular option)
+	fieldOptionsPaths := fieldOptionsTrie{}
 	for i, location := range sourceCodeInfo.Location {
-		if _, ok := paths[GetPathKey(location.Path)]; !ok {
+		path := location.Path
+		if isPathForFieldOptions(path) {
+			fieldOptionsPaths.insert(path, i)
+		}
+		if _, ok := paths[GetPathKey(path)]; !ok {
+			if isPathForSomeFieldOption(path) {
```

Comment:
`isPathForFieldOption` could return a second bool to indicate whether it's a path to the options message directly or to a descendant (or maybe return a new enum type instead of two bools), so that you don't have to run the matcher 2x.